sbt-plugin/src/main/scala/akka/grpc/sbt/AkkaGrpcPlugin.scala

```diff
@@ -117,41 +136,35 @@ object AkkaGrpcPlugin extends AutoPlugin {
-  def generatorsFor(stubs: Seq[AkkaGrpc.GeneratedSource], languages: Seq[AkkaGrpc.Language], scalaBinaryVersion: ScalaBinaryVersion, logger: GenLogger): Seq[protocbridge.Generator] = {
+  def generatorsFor(stubs: Seq[AkkaGrpc.GeneratedSource], languages: Seq[AkkaGrpc.Language], options: Seq[String], scalaBinaryVersion: ScalaBinaryVersion, logger: GenLogger): Seq[protocbridge.Generator] = {
+    import AkkaGrpc._
     // these two are the model/message (protoc) generators
     def ScalaGenerator: protocbridge.Generator = protocbridge.JvmGenerator("scala", ScalaPbCodeGenerator)
     // we have a default flat_package, but that doesn't play with the java generator (it fails)
     def JavaGenerator: protocbridge.Generator = PB.gens.java
 
-    stubs match {
-      case Seq(_, _) =>
-        languages match {
-          case Seq(_, _) => Seq(ScalaGenerator, toGenerator(ScalaBothCodeGenerator, scalaBinaryVersion, logger), JavaGenerator, toGenerator(JavaBothCodeGenerator, scalaBinaryVersion, logger))
-          case Seq(AkkaGrpc.Scala) => Seq(ScalaGenerator, toGenerator(ScalaBothCodeGenerator, scalaBinaryVersion, logger))
-          case Seq(AkkaGrpc.Java) => Seq(JavaGenerator, toGenerator(JavaBothCodeGenerator, scalaBinaryVersion, logger))
-        }
-      case Seq(AkkaGrpc.Client) =>
-        languages match {
-          case Seq(_, _) => Seq(ScalaGenerator, toGenerator(ScalaClientCodeGenerator, scalaBinaryVersion, logger), JavaGenerator, toGenerator(JavaClientCodeGenerator, scalaBinaryVersion, logger))
-          case Seq(AkkaGrpc.Scala) => Seq(ScalaGenerator, toGenerator(ScalaClientCodeGenerator, scalaBinaryVersion, logger))
-          case Seq(AkkaGrpc.Java) => Seq(JavaGenerator, toGenerator(JavaClientCodeGenerator, scalaBinaryVersion, logger))
-        }
-      case Seq(AkkaGrpc.Server) =>
-        languages match {
-          case Seq(_, _) => Seq(ScalaGenerator, toGenerator(ScalaServerCodeGenerator, scalaBinaryVersion, logger), JavaGenerator, toGenerator(JavaServerCodeGenerator, scalaBinaryVersion, logger))
-          case Seq(AkkaGrpc.Scala) => Seq(ScalaGenerator, toGenerator(ScalaServerCodeGenerator, scalaBinaryVersion, logger))
-          case Seq(AkkaGrpc.Java) => Seq(JavaGenerator, toGenerator(JavaServerCodeGenerator, scalaBinaryVersion, logger))
-        }
-    }
+    val serverPowerApis = options.contains(GeneratorOption.ServerPowerApis.setting)
+
+    (for {
+      stub <- stubs
+      language <- languages
+    } yield (stub, language) match {
+      case (_, Java) => logger.error("Livongo: java code generators not yet available"); Seq.empty
+      case (Client, Scala) => Seq(ScalaGenerator, toGenerator(ScalaClientCodeGenerator, scalaBinaryVersion, logger))
+      case (PlayClient, Scala) => Seq(ScalaGenerator, toGenerator(PlayScalaClientCodeGenerator, scalaBinaryVersion, logger))
+      case (Server, Scala) => Seq(ScalaGenerator, toGenerator(ScalaServerCodeGenerator(serverPowerApis), scalaBinaryVersion, logger))
```

Comment:
You're right, that was broken. But when I looked at how to update this logic to handle using that correctly with all the possible combinations (Client, PlayClient) x (Server, PlayServer) it looked to be quite ugly. A cleaner approach I implemented was creating a standalone `ScalaTraitCodeGenerator` to generate the service trait, and changing the client and server code generators to only generate their files and not also the trait. Doing so eliminated the need for `ScalaBothCodeGenerator` and allowed this logic in `AkkaGrpcPlugin` to stay simple.

Reply:
I agree the current approach there is relatively ugly, and it would be worth investigating how it can be improved. I was a bit afraid that doing this in this PR might make it bigger and slow it down, which is why I didn't propose doing it here - if you don't mind then doing it here is OK by me.